File: src/main/java/cn/mobiledaily/service/ExhibitionServiceImpl.java
package cn.mobiledaily.service;

import cn.mobiledaily.domain.Exhibition;
import cn.mobiledaily.domain.News;
import cn.mobiledaily.domain.mobile.ExhibitionContent;
import cn.mobiledaily.repository.ExhibitionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Service("exhibitionService")
@Transactional(readOnly = true)
public class ExhibitionServiceImpl implements ExhibitionService {
    @Autowired
    private ExhibitionRepository exhibitionRepository;

    @Override
    @Transactional
    public void save(Exhibition exhibition) {
        exhibitionRepository.save(exhibition);
    }

    @Override
    @Transactional
    public void save(ExhibitionContent content) {
        Exhibition exhibition = content.getExhibition();
        exhibition.setUpdatedAt(new Date());
        exhibitionRepository.save(content);
        exhibitionRepository.save(exhibition);
    }

    @Override
    public List<Exhibition> findAll() {
        return exhibitionRepository.findAll();
    }

    @Override
    public Exhibition findByCode(String code) {
        return exhibitionRepository.findByCode(code);
    }

    @Override
    public Exhibition findById(String id) {
        return exhibitionRepository.findById(id);
    }

    @Override
    @Transactional
    public void delete(Exhibition exhibition) {
        exhibitionRepository.delete(exhibition);
    }

    @Override
    public <T extends ExhibitionContent> List<T> findContents(String exhibitionCode, Class<T> type, Sort sort) {
        try {
            Exhibition exhibition = findByCode(exhibitionCode);
            return exhibitionRepository.findContents(exhibition, type, sort);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T extends ExhibitionContent> List<T> findContents(String exhibitionCode, Class<T> type) {
        return findContents(exhibitionCode, type, null);
    }

    @Override
    public <T extends ExhibitionContent> T findContentById(String exhibitionCode, String id, Class<T> type) {
        try {
            Exhibition exhibition = findByCode(exhibitionCode);
            return exhibitionRepository.findContentById(exhibition, id, type);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<News> findNews(String exhibitionCode, long from, int size) {
        try {
            Exhibition exhibition = findByCode(exhibitionCode);
            return exhibitionRepository.findNews(exhibition, from, size);
        } catch (Exception e) {
            return new ArrayList<>(0);
        }
    }

    @Override
    @Transactional
    public void delete(ExhibitionContent content) {
        exhibitionRepository.delete(content);
    }
}

File: src/main/scala/cn/mobiledaily/repository/ExhibitionServiceImpl.scala
package cn.mobiledaily.repository

import cn.mobiledaily.service.ExhibitionService
import cn.mobiledaily.domain.mobile.ExhibitionContent
import org.springframework.data.domain.Sort
import java.util
import cn.mobiledaily.domain.Exhibition
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.beans.factory.annotation.Autowired
import java.util.Date

@Service("exhibitionService")
@Transactional(readOnly = true)
class ExhibitionServiceImpl extends ExhibitionService{
    @Autowired
    var exhibitionRepository: ExhibitionRepository = null

    def save(exhibition: Exhibition) {exhibitionRepository.save(exhibition)}

    def save(content: ExhibitionContent) {
        val exhibition = content.getExhibition
        exhibition.updatedAt = new Date()
        exhibitionRepository.save(content)
        exhibitionRepository.save(exhibition)
    }

    def findAll(): util.List[Exhibition] = exhibitionRepository.findAll()

    def findByCode(code: String): Exhibition = exhibitionRepository.findByCode(code)

    def findById(id: String): Exhibition = exhibitionRepository.findById(id)

    def delete(exhibition: Exhibition) {exhibitionRepository.delete(exhibition)}

    def findContents[T <: ExhibitionContent](exhibitionCode: String, `type`: Class[T]): util.List[T] = findContents(exhibitionCode, `type`, null)

    def findContents[T <: ExhibitionContent](exhibitionCode: String, `type`: Class[T], sort: Sort): util.List[T] = {
        val exhibition = findByCode(exhibitionCode)
        exhibitionRepository.findContents(exhibition, `type`, sort)
    }

    def findContentById[T <: ExhibitionContent](exhibitionCode: String, id: String, `type`: Class[T]): T = {
        try {
            val exhibition = findByCode(exhibitionCode)
            exhibitionRepository.findContentById(exhibition, id, `type`)
        } catch {
            case e: Exception => throw new RuntimeException(e)
        }
    }

    def delete[T <: ExhibitionContent](content: ExhibitionContent) {exhibitionRepository.delete(content)}
}
